### PR TITLE
Make interface between Wasm guests and the host synchronous

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Wasm applications were previously expected to be asynchronous. The initial goals were to allow more concurrency, especially for storage operations. However, that led to increased complexity, higher fuel consumption by contracts, and less debugability, and concurrent storage access is still achievable with a simpler interface.

## Proposal

Replace the asynchronous system APIs with "promise" interfaces, which aren't polled but are waited upon after they are created. This by itself already ensures determinism in the contracts, but the `HostFutureQueue` will be removed in a separate PR that also uses a `RwLock` to avoid data-races with the new promise API.

Replace the guest Wasm entrypoints with synchronous functions. The SDK is still asynchronous, which means the entrypoints will simply poll them repeatedly until they complete. This is safe because there are no more system APIs that are asynchronous, so there are no more waiting points for polls.

## Test Plan

This refactors the Wasm application interfaces, and the integration tests cover them.

## Release Plan

This changes the API for applications, so:
- Need to bump the major/minor version number in the next release of the crates.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-sdk/src/service/system_api.rs [hunk 1]
```diff
@@ -38,17 +29,15 @@ where
 
 /// Loads the application state (and locks it for writes).
 pub(crate) async fn lock_and_load_view<State: View<ViewStorageContext>>() -> State {
-    let future = wit::Lock::new();
-    future::poll_fn(|_context| -> Poll<Result<(), ViewError>> { future.poll().into() })
-        .await
-        .expect("Failed to lock application state");
+    let promise = wit::Lock::new();
+    promise.wait().expect("Failed to lock application state");
```
Comment (ma2bd) [inline linera-sdk/src/service/system_api.rs:33]: This seems to be waiting immediately after creating the promise (no poll).

Comment (jvff) [reply linera-sdk/src/service/system_api.rs]: Fixed, thanks! There was a minor unexpected issue though, one test called the API using `now_or_never`, which now would return `None`. I fixed it by blocking on the future instead.

File: linera-execution/src/wasm/common.rs [hunk 2]
```diff
@@ -34,165 +35,60 @@ pub trait ApplicationRuntimeContext: Sized {
...
+    ) -> Result<Result<RawExecutionResult<Vec<u8>>, String>, Self::Error>;
 
-    /// Creates a new future for the user contract to handle a call from another contract.
-    fn handle_application_call_new(
+    /// Handles a call from another contract.
+    fn handle_application_call(
         &self,
         store: &mut Self::Store,
         context: CalleeContext,
         argument: Vec<u8>,
```
Comment (ma2bd) [inline linera-execution/src/wasm/common.rs:67]: Would a `&[u8]` work now, here and in the other places?

Comment (ma2bd) [reply linera-execution/src/wasm/common.rs:67]: Same for the `*Context` arguments

File: linera-execution/src/wasm/common.rs [hunk 3]
```diff
@@ -352,8 +274,59 @@ where
...
+        })
+    }
+}
+
+impl<A> WasmRuntimeContext<A>
+where
+    A: ApplicationRuntimeContext + Send + Unpin + 'static,
+{
+    /// Spawns a thread to execute a Wasm guest.
+    fn run_wasm_guest<T>(
```
Comment (ma2bd) [inline linera-execution/src/wasm/common.rs:296]: This is nice!

File: linera-execution/src/wasm/common.rs [hunk 4]
```diff
@@ -352,8 +274,59 @@ where
...
+        T: Send + 'static,
+    {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        // TODO(#925): Replace the host future queue with a read-write-lock.
+        if let Some(future_queue) = self.future_queue.take() {
+            tokio::spawn(future_queue.collect::<()>());
+        }
+
+        thread::spawn(move || {
```
Comment (ma2bd) [inline linera-execution/src/wasm/common.rs:312]: Do we have a task for merging the threads of a transaction?

File: linera-execution/tests/wasm.rs [hunk 5]
```diff
@@ -25,10 +25,10 @@ use test_case::test_case;
 /// called correctly and consume the expected amount of fuel.
 ///
 /// To update the bytecode files, run `linera-execution/update_wasm_fixtures.sh`.
-#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::Wasmer, 30_463; "wasmer"))]
-#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::WasmerWithSanitizer, 30_844; "wasmer_with_sanitizer"))]
-#[cfg_attr(feature = "wasmtime", test_case(WasmRuntime::Wasmtime, 30_844; "wasmtime"))]
-#[cfg_attr(feature = "wasmtime", test_case(WasmRuntime::WasmtimeWithSanitizer, 30_844; "wasmtime_with_sanitizer"))]
+#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::Wasmer, 29_327; "wasmer"))]
```
Comment (ma2bd) [inline linera-execution/tests/wasm.rs:28]: Looks like this is going in the right direction!

File: linera-sdk/src/util.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Module with helper types and functions used by the SDK.
```
Comment (ma2bd) [inline linera-sdk/src/util.rs:4]: Nice! Perhaps include a couple of unit tests in the same file?

**Review by ma2bd — APPROVED** [review]
The code looks much simpler!

File: linera-sdk/src/util.rs (hunk 6 above)
Comment (jvff) [reply linera-sdk/src/util.rs:4]: Good idea! Done.

File: linera-execution/src/wasm/common.rs (hunk 4 above)
Comment (jvff) [reply linera-execution/src/wasm/common.rs:313]: Yes, I added a link in the code. Thanks!

File: linera-execution/src/wasm/common.rs (hunk 2 above)
Comment (jvff) [reply linera-execution/src/wasm/common.rs:67]: No, unfortunately. The parameters are still sent to a different thread (inside `run_wasm_guest`), so they need to be `Send + 'static`.

**Pull request merged**